topup/handler.go

```diff
@@ -0,0 +1,161 @@
+	hmTypes "github.com/maticnetwork/heimdall/types"
+)
+
+// NewHandler returns a handler for "bank" type messages.
+func NewHandler(k Keeper, contractCaller helper.IContractCaller) sdk.Handler {
+	return func(ctx sdk.Context, msg sdk.Msg) sdk.Result {
+		ctx = ctx.WithEventManager(sdk.NewEventManager())
+
+		switch msg := msg.(type) {
+		case types.MsgTopup:
+			return handleMsgTopup(ctx, k, msg, contractCaller)
+		case types.MsgWithdrawFee:
+			return handleMsgWithdrawFee(ctx, k, msg)
+		default:
+			errMsg := "Unrecognized bank Msg type: %s" + msg.Type()
+			return sdk.ErrUnknownRequest(errMsg).Result()
+		}
+	}
+}
+
+// Handle MsgMintFeeToken
+func handleMsgTopup(ctx sdk.Context, k Keeper, msg types.MsgTopup, contractCaller helper.IContractCaller) sdk.Result {
+	if !k.bk.GetSendEnabled(ctx) {
+		return types.ErrSendDisabled(k.Codespace()).Result()
+	}
+
+	// get main tx receipt
+	receipt, err := contractCaller.GetConfirmedTxReceipt(msg.TxHash.EthHash())
+	if err != nil || receipt == nil {
+		return hmCommon.ErrWaitForConfirmation(k.Codespace()).Result()
+	}
+
+	// get event log for topup
+	eventLog, err := contractCaller.DecodeValidatorTopupFeesEvent(receipt, msg.LogIndex)
+	if err != nil || eventLog == nil {
+		k.Logger(ctx).Error("Error fetching log from txhash")
+		return hmCommon.ErrInvalidMsg(k.Codespace(), "Unable to fetch logs for txHash").Result()
+	}
+
+	if eventLog.ValidatorId.Uint64() != msg.ID.Uint64() {
+		k.Logger(ctx).Error("ID in message doesn't match id in logs", "MsgID", msg.ID, "IdFromTx", eventLog.ValidatorId)
+		return hmCommon.ErrInvalidMsg(k.Codespace(), "Invalid txhash and id don't match. Id from tx hash is %v", eventLog.ValidatorId.Uint64()).Result()
+	}
+
+	// use event log signer
+	signer := hmTypes.BytesToHeimdallAddress(eventLog.Signer.Bytes())
+	// if validator exists use siger from local state
+	validator, found := k.vm.GetValidatorFromValID(ctx, msg.ID)
+	if found {
+		signer = validator.Signer
+	}
+
+	// validator topup
+	topupObject, err := k.GetValidatorTopup(ctx, signer)
+	if err != nil {
+		return types.ErrNoValidatorTopup(k.Codespace()).Result()
+	}
+
+	// create topup object
+	if topupObject == nil {
+		topupObject = &types.ValidatorTopup{
+			ID:          msg.ID,
+			TotalTopups: hmTypes.Coins{hmTypes.Coin{Denom: authTypes.FeeToken, Amount: hmTypes.NewInt(0)}},
+		}
+	}
+
+	// create topup amount
+	topupAmount := hmTypes.Coins{hmTypes.Coin{Denom: authTypes.FeeToken, Amount: hmTypes.NewIntFromBigInt(eventLog.Fee)}}
+
+	// sequence id
+	sequence := (receipt.BlockNumber.Uint64() * hmTypes.DefaultLogIndexUnit) + msg.LogIndex
+
+	// check if incoming tx already exists
+	if k.HasTopupSequence(ctx, sequence) {
+		k.Logger(ctx).Error("Older invalid tx found")
+		return hmCommon.ErrOldTx(k.Codespace()).Result()
+	}
+
+	// add total topups amount
+	topupObject.TotalTopups = topupObject.TotalTopups.Add(topupAmount)
+
+	// increase coins in account
+	if _, ec := k.bk.AddCoins(ctx, signer, topupAmount); ec != nil {
```

Comment:
Please use more descriptive name instead of `ec`

Reply:
umm, not sure
Can you please suggest? :grin:

Reply:
`err`?

Reply:
yep